File: leetcode/0347-top-k-frequent-elements/solution.py
from typing import List
from collections import Counter


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counts = Counter(nums)
        n = len(nums)

        buckets = [[] for _ in range(n + 1)]
        for value, freq in counts.items():
            buckets[freq].append(value)

        result = []
        for freq in range(n, 0, -1):
            for value in buckets[freq]:
                result.append(value)
                if len(result) == k:
                    return result

        return result

File: leetcode/0347-top-k-frequent-elements/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> counts;
        for (int x : nums) counts[x]++;

        int n = nums.size();
        vector<vector<int>> buckets(n + 1);
        for (auto& [value, freq] : counts) {
            buckets[freq].push_back(value);
        }

        vector<int> result;
        for (int freq = n; freq >= 1 && (int)result.size() < k; freq--) {
            for (int value : buckets[freq]) {
                result.push_back(value);
                if ((int)result.size() == k) break;
            }
        }

        return result;
    }
};
